IdentityCore/src/cache/mac/MSIDMacCredentialStorageItem.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+
+#import <Foundation/Foundation.h>
+#import "MSIDCredentialCacheItem.h"
+#import "MSIDDefaultCredentialCacheKey.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDMacCredentialStorageItem : NSObject <MSIDJsonSerializable>
+
+- (void)storeCredential:(MSIDCredentialCacheItem *)credential forKey:(MSIDDefaultCredentialCacheKey *)key;
+
+/*
+ This api is thread safe only if an immutable object is passed as parameter.
```

Comment:
What do you mean by "immutable object"? Do we have a mutable/immutable version of this class like (NSMutableArray vs NSArray)? If you just mean that developer should verify that passed MSIDMacCredentialStorageItem is not being modified during the merge operation, isn't it error prone? I would prefer this method to take care of thread safety by coping received storage item before merging it.

Reply:
We had a discussion about this in the morning and came up to a conclusion of not copying the passed in storage item as it will be an expensive operation. In the current code, storage item is a local object reference which is passed to the merge credential and is destroyed after the merge completes so it is thread safe. Since this is not a public api exposed to the developer, we decided to add a comment for immutability for internal usage only and decided to avoid a copy.

Reply:
I agree that from the API perspective it might be unclear that this won't be thread safe internally.
However, there was a good point that making it thread safe will possibly have an additional performance hit (this cache object might possibly grow big and copying it entirely will have a hit), although this API will never be used from multiple threads in actual implementation. We agreed that as long as it's super clear that this API is not thread safe, we can proceed.